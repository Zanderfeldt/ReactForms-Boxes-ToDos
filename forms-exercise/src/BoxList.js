import React, { useState } from "react";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";

const BoxList = () => {
  const [boxes, setBoxes] = useState([]);
  const addBox = (newBox) => {
    setBoxes(boxes => [...boxes, {...newBox}])
  };
  const removeBox = id => {
    setBoxes(boxes => boxes.filter(box => box.id !== id));
  };
  return (
    <div>
      <h3>Boxes!</h3>
      <NewBoxForm addBox={addBox}/>
      <div>
        {boxes.map(({id, color, height, width}) => (
        <Box
          key={id}
          id={id}
          color={color} 
          width={width} 
          height={height}
          removeBox={removeBox}/>
        ))}
      </div>
    </div>
  )
}

export default BoxList