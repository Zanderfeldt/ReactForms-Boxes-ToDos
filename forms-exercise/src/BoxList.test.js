import React from "react";
import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

it("renders without crashing", () => {
  render(<BoxList />);
});

it("matches snapshot", () => {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});

it("should add new box", () => {
  const { queryByText, getByLabelText, getByText, getAllByDisplayValue } = render(<BoxList />);

  expect(queryByText("X")).not.toBeInTheDocument();
  fireEvent.change(getByLabelText("Height"), { target: { value: '5' }});
  fireEvent.change(getByLabelText("Width"), { target: { value: '5' }});
  fireEvent.change(getByLabelText("Color"), { target: { value: 'red'}});
  const button = getByText("Create Box");
  fireEvent.click(button);

  const removeButton = getByText("X");
  expect(removeButton).toBeInTheDocument();
  expect(removeButton.previousSibling).toHaveStyle(`
    width: 5em;
    height: 5em;
    background-color: red;
  `);
  expect(getAllByDisplayValue("")).toHaveLength(3);
});

it("should remove a box", () => {
  const { getByLabelText, getByText, } = render(<BoxList />);

  fireEvent.change(getByLabelText("Height"), { target: { value: '5' }});
  fireEvent.change(getByLabelText("Width"), { target: { value: '5' }});
  fireEvent.change(getByLabelText("Color"), { target: { value: 'red'}});
  const button = getByText("Create Box");
  fireEvent.click(button);

  const removeButton = getByText("X");
  fireEvent.click(removeButton);
  expect(removeButton).not.toBeInTheDocument();
});