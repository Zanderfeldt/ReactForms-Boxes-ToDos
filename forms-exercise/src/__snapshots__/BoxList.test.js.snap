// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h3>
      Boxes!
    </h3>
    <form>
      <label
        for="color"
      >
        Color
      </label>
      <input
        id="color"
        name="color"
        placeholder="Box Color"
        type="text"
        value=""
      />
      <label
        for="height"
      >
        Height
      </label>
      <input
        id="height"
        name="height"
        placeholder="Box Height"
        type="text"
        value=""
      />
      <label
        for="width"
      >
        Width
      </label>
      <input
        id="width"
        name="width"
        placeholder="Box Width"
        type="text"
        value=""
      />
      <button>
        Create Box
      </button>
    </form>
    <div />
  </div>
</DocumentFragment>
`;
